rust/agents/relayer/src/merkle_tree/builder.rs

```diff
@@ -81,54 +78,43 @@ impl MerkleTreeBuilder {
     #[instrument(err, skip(self), level="debug", fields(prover_latest_index=self.count()-1))]
     pub fn get_proof(
         &self,
-        leaf_index: u32,
+        message_nonce: u32,
         root_index: u32,
-    ) -> Result<Proof, MerkleTreeBuilderError> {
+    ) -> Result<Option<Proof>, MerkleTreeBuilderError> {
+        let Some(message_id) = self
+            .db
+            .retrieve_message_id_by_nonce(&message_nonce)?
+        else {
+            return Ok(None);
+        };
+        let Some(leaf_index) = self
+            .db
+            .retrieve_merkle_leaf_index_by_message_id(&message_id)?
+        else {
+            return Ok(None);
+        };
         self.prover
             .prove_against_previous(leaf_index as usize, root_index as usize)
+            .map(Option::from)
             .map_err(Into::into)
     }
 
-    fn ingest_nonce(&mut self, nonce: u32) -> Result<(), MerkleTreeBuilderError> {
-        match self.db.retrieve_message_id_by_nonce(&nonce) {
-            Ok(Some(leaf)) => {
-                debug!(nonce, "Ingesting leaf");
-                self.prover.ingest(leaf).expect("!tree full");
-                self.incremental.ingest(leaf);
-                assert_eq!(self.prover.root(), self.incremental.root());
-                Ok(())
-            }
-            Ok(None) => {
-                error!("We should not arrive here");
-                Err(MerkleTreeBuilderError::UnavailableNonce { nonce })
-            }
-            Err(e) => Err(e.into()),
-        }
-    }
-
     pub fn count(&self) -> u32 {
         self.prover.count() as u32
     }
 
-    #[instrument(err, skip(self), level = "debug")]
-    pub async fn update_to_index(&mut self, index: u32) -> Result<(), MerkleTreeBuilderError> {
-        if index >= self.count() {
-            let starting_index = self.prover.count() as u32;
-            for i in starting_index..=index {
-                self.db.wait_for_message_nonce(i).await?;
-                self.ingest_nonce(i)?;
-            }
-
-            let prover_root = self.prover.root();
-            let incremental_root = self.incremental.root();
-            if prover_root != incremental_root {
-                return Err(MerkleTreeBuilderError::MismatchedRoots {
-                    prover_root,
-                    incremental_root,
-                });
-            }
+    pub async fn ingest_message_id(&mut self, message_id: H256) -> Result<()> {
+        const CTX: &str = "When ingesting message id";
+        debug!(?message_id, "Ingesting leaf");
+        self.prover.ingest(message_id).expect("!tree full");
```

Comment:
nit: would probably remove the `!`, if this panics then I'd read this as meaning "tree not full" when it is in fact full